### PR TITLE
[TOOLS-2964] Scenario - Add week

### PR DESCRIPTION
**Pull request by jhonypg**

### CODE REVIEW

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java [hunk 1]
```diff
@@ -129,6 +137,19 @@ protected ScenarioModel withStartSimulationDate(LocalDate weekDate) {
                 .map(WeekParametersModel::toDetailsDto);
     }
 
+    @Override
+    public List<WeekDateDto> availableWeeksToBeCreated() {
+        List<WeekDateDto> weeksThatCanBeCreated = project.weeksThatCanBeCreated();
+        currentScenarioWeeks()
+                .map(Map.Entry::getKey)
+                .forEach(scenarioWeekDate -> {
```
Comment (objective-solutions-user) [inline dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java]: MINOR Remove useless curly braces around statement [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java [hunk 2]
```diff
@@ -129,6 +137,19 @@ protected ScenarioModel withStartSimulationDate(LocalDate weekDate) {
...
+                    weeksThatCanBeCreated.removeIf(weekThatToCreate -> {
```
Comment (objective-solutions-user) [inline dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java]: MINOR Remove useless curly braces around statement and then remove useless return keyword [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

File: dashboard-scenario/src/test/java/objective/taskboard/scenario/model/StandardScenarioModelTest.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package objective.taskboard.scenario.model;
+
+public class StandardScenarioModelTest {
```
Comment (objective-solutions-user) [inline dashboard-scenario/src/test/java/objective/taskboard/scenario/model/StandardScenarioModelTest.java]: BLOCKER Add some tests to this class. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2187)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 3 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java [hunk 4]
```diff
@@ -129,6 +137,16 @@ protected ScenarioModel withStartSimulationDate(LocalDate weekDate) {
                 .map(WeekParametersModel::toDetailsDto);
     }
 
+    @Override
+    public List<WeekDateDto> availableWeeksToBeCreated() {
+        List<WeekDateDto> weeksThatCanBeCreated = project.weeksThatCanBeCreated();
+        currentScenarioWeeks()
+                .map(Map.Entry::getKey)
+                .forEach(scenarioWeekDate ->
```
Comment (objective-solutions-user) [inline dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java]: MINOR Remove useless curly braces around statement [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java [hunk 5]
```diff
@@ -129,6 +137,16 @@ protected ScenarioModel withStartSimulationDate(LocalDate weekDate) {
...
+                    weeksThatCanBeCreated.removeIf(weekThatToCreate -> scenarioWeekDate.isEqual(weekThatToCreate.weekDate)));
```
Comment (objective-solutions-user) [inline dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java]: MINOR Remove useless curly braces around statement and then remove useless return keyword [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1602)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 1 issue
* ![BLOCKER][BLOCKER] 1 blocker

Watch the comments in this conversation to review them.

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'

File: dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-form.component.ts [hunk 6]
```diff
@@ -20,21 +20,27 @@ import { Observable } from 'rxjs';
...
 
     @ViewChild('modal') scenarioModal: ModalComponent;
     @ViewChild('scenarioForm') scenarioForm: NgForm;
+    @ViewChild('scrollArea') scrollArea: ElementRef;
     @ViewChildren('input') inputs: QueryList<ElementRef>;
     @ViewChildren(ScenarioDetailsComponent) scenarioDetailsComponent: QueryList<ScenarioDetailsComponent>;
 
     constructor(
         private route: ActivatedRoute,
+        @Inject(DATE_INPUT_DISPLAY_FORMAT) public dateDisplayFormat: string,
```
Comment (jytakano) [inline dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-form.component.ts]: Not used
Remember to remove imports too

File: dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-details/scenario-details.component.ts [hunk 7]
```diff
@@ -50,6 +68,10 @@ export class ScenarioDetailsComponent implements OnInit {
     }
 
     restore(event: Event) {
+        if (this.scenarioDatesHandler)
```
Comment (jytakano) [inline dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-details/scenario-details.component.ts]:
```suggestion
        if (this.scenarioDatesHandler && this.selectedDate)
```

File: dashboard-scenario/src/test/java/objective/taskboard/scenario/model/ProjectModelTest.java [hunk 8]
```diff
@@ -93,10 +99,130 @@ public void pastLogbooks_givenStartedProjectWithoutLogbooks_thenGenerateItAndRet
...
+        .whenRun(() -> dateRange = subject.futureWeeks())
+
+        .thenAssert()
+            .dateRangesAreEquals(dateRange, rangeOf(weekOf("2018-01-01"), weekEndDateOf("2018-01-01")));
+    }
+
+    @Test
+    public void weeksThatCanBeCreated_givenCurrentDateBeingBeforeDeliveryDate_thenReturnStartingFromDeliveryDateToProjectionLimitDate() {
+        given()
+            .projectStartDate("2017-02-20")
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply dashboard-scenario/src/test/java/objective/taskboard/scenario/model/ProjectModelTest.java]: Yes. StartDate is being used to calculate projectionLimitDate()

File: dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java [hunk 9]
```diff
@@ -208,21 +247,31 @@ private void added(UpdateScenarioDetailsDto data) {
...
+        return entity.getWeekParameters().entrySet().stream()
                 .filter(entry -> {
                     LocalDate week = entry.getKey();
                     return week.equals(currentWeek) || week.isAfter(currentWeek);
                 });
+    }
 
-        return Stream.concat(lastLogbookWeek, scenarioWeeks)
-                .collect(toTreeMap(Map.Entry::getKey, Map.Entry::getValue));
+    private Optional<LocalDate> lastScenarioWeek() {
```
[the remark this answers is not in the document]
Comment (jhonypg) [reply dashboard-scenario/src/main/java/objective/taskboard/scenario/model/StandardScenarioModel.java]: Done.

File: dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-form.component.ts (hunk 6 above)
Comment (jhonypg) [reply dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-form.component.ts]: Done

File: dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-details/scenario-details.component.ts (hunk 7 above)
Comment (jhonypg) [reply dashboard-scenario/web-app/src/app/scenario/scenario/form/scenario-details/scenario-details.component.ts]: Done.

**Pull request merged**